apps/speed.c

```diff
@@ -298,19 +304,20 @@ static double Time_F(int s)
 
     if (s == START) {
         schlock = 0;
-        thr = CreateThread(NULL, 4096, sleepy, NULL, 0, NULL);
+        stopevt = CreateEvent(NULL, TRUE, FALSE, NULL);
+        thr = (HANDLE)_beginthreadex(NULL, 4096, sleepy, NULL, 0, NULL);
         if (thr == NULL) {
             DWORD err = GetLastError();
-            BIO_printf(bio_err, "unable to CreateThread (%lu)", err);
+            BIO_printf(bio_err, "_beginthreadex failed (%lu)", err);
             ExitProcess(err);
         }
         while (!schlock)
             Sleep(0);           /* scheduler spinlock */
         ret = app_tminterval(s, usertime);
     } else {
         ret = app_tminterval(s, usertime);
-        if (run)
-            TerminateThread(thr, 0);
+        if (stopevt != NULL)    /* if stopevt is NULL, thread has timed out */
+            SetEvent(stopevt);  /* signal stopevt to allow thread to return */
```

Comment:
Formally speaking there is race condition here. In sense that you can end up calling SetEvent on closed handler. And by the way, it should also be declared volatile.